grove-node-server-utils/auth-helper.js

```diff
@@ -290,18 +297,64 @@ function getAuthorization(session, reqMethod, reqPath, authOptions) {
   return d.promise;
 }
 
-// TODO, but looks broken at the moment (authenticator isn't used)
-// function expiresAt(session) {
-//   var authenticator = getAuthenticator(
-//     session,
-//     session.passport.user,
-//     defaultOptions.authHost,
-//     defaultOptions.authPort
-//   );
-//   var d = new Date();
-//   d.setTime(d.getTime() + expirationTime);
-//   return d;
-// }
+function genNonce(b) {
+  var c = [],
+    e = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789',
+    a = e.length;
+  for (var d = 0; d < b; ++d) {
+    c.push(e[(Math.random() * a) | 0]);
+  }
+  return c.join('');
+}
+
+function genDigestSecret(username, password, realm) {
+  return md5Hex(username + ':' + realm + ':' + password);
+}
+
+function HttpDigestHeader(params, SECRET) {
+  /*
+    { 
+        nc: optional,
+        cnonce: optional,
+        username: username,
+        method: reqMethod,
+        uri: reqPath,
+        realm: 'public',
+        qop: 'auth',
+        nonce: '379fb206fbe88e:urD2gEqZJ7ApwKmJHDz1Gw==',
+        opaque: '9d9c655418122b53' }
+    */
+  let nonce = params.nonce; // unique string value specified by the server
+  let realm = params.realm; // authentication realm, defaults to "IsmuserAPI"
+
+  let qop = params.qop; // list of quality of protection values supported by the server (auth | auth-int); auth set.
+
+  let USER = params.username;
+  let method = params.method;
+  let URI = params.uri;
+
+  let cnonce = params.cnonce; // opaque random string value provided by the client
+  let nc = params.nc; // count of the number of requests (including the current request) that the client has sent with the nonce value
+
+  //then use previous and increment nonce
+  if (params.cnonce == null || params.nc == null) {
+    //generate
+    cnonce = genNonce(10);
+    nc = 0;
+  }
+
+  //SECRET = md5Hex(user:realm:pass)
+
+  // ha2 = md5Hex(method:uri)
+  let HA2 = md5Hex(method + ':' + URI);
+
+  // response = md5Hex((SECRET + ":" + nonce + ":" + nc + ":" + cnonce + ":" + qop + ":" + HA2)
+  let response = md5Hex(`${SECRET}:${nonce}:${nc}:${cnonce}:${qop}:${HA2}`);
+
+  const header = `Digest username="${USER}", realm="${realm}", nonce="${nonce}", uri="${URI}", cnonce="${cnonce}", nc="${nc}", qop="${qop}", response="${response}"`;
+
+  return header;
+}
```

Comment:
Wondering, where did you get this from? Maybe add a comment with the source. Is there perhaps a lib that can do this for us, rather than having to write this out ourselves?

Reply:
I drew inspiration from these 2 examples
https://dev.to/leolanchas/how-to-implement-http-digest-rfc-2617-in-javascript-in-node-using-passportjs-17l8
https://gist.github.com/funkatron/949952/11c11ef47f8dab54722ee20dc33372b7417579a6

I also tried pretty hard to find a library but the problem I had and the issue with the existing www-authenticate library is that I don't want to create/recreate the authenticator from a stored password, Instead the most secure method seems to be to create the hashed secret once, store that in the session and regenerate authorization header each time using the secret. 

This means we never need to store or keep the password in memory other than the initial login request.